src/bsp/include/bsp/ui_impl.h

```diff
@@ -0,0 +1,14 @@
+#ifndef __UI_IMPL_H_
+#define __UI_IMPL_H_
+
+#include "bsp/ui.h"
+
+class UIImpl : public UI {
```

Comment:
Are we naming the concrete class with a `Impl` suffix?
Wouldn't it be better to add a I to the interface instead?

Reply:
Yea I thought about this, what does the I prefix actually brings? The user of the UI library doesn't need to know that it's an interface, only the maintainer needs to know that. We need to add interface to everything anyway since we have unit testing and the mocking requires it. The I prefix is only a standard in C#, for exemple in Java the List, Collection, Iteratable are interfaces but are not prefixed with I.

I knew this would come up in the PR. @stlaperriere @t-constantin @g-lesssard what do you guys think? Let's just pick a standard and go on with it, as long as it's consistent. The I prefix is fine if the other agrees, I'm just not sure what benefits it brings.

See this post for a bit more info:
 https://softwareengineering.stackexchange.com/questions/117348/should-interface-names-begin-with-an-i-prefix

Reply:
I don't mind not using `I`, but if not using it means we have `Impl` on concrete classes I'd rather have the `I`

Reply:
The main reason is to differentiate from the class implementation with some sort of consistency. Either we add something to the implementation of the interface definition. The same interface could be implemented by multiple classes (IState for example) and is, therefore, more efficient. Also, a class does not require to implement an interface, so a trailing Impl would be useless, and not having it might be confusing to differentiate it from interfaces. 

For the part of the user knowing that the class is an interface, why is it a problem if the user knows? On the contrary, I would advise that it should be known if it needs to be mocked for unit testing on his side. By design, the UI class exposed to the user should be an interface, so having a prefix to the class only confirms it for the user.

TLDR: a prefix or suffix should be added to differentiate an interface from its implementation and it seems more efficient to add it to the interface. It would also add clarity for external users, which might not be needed but wouldn't be harmful either.

Reply:
My many decades of writing oo code in Java (sike) have never brought me to having to name my interfaces the same way as I name the implementations. So I never really had this problem. But then again, in Java the interfaces are specifically declared as such, and not as classes that have their virtual methods `=0`. And I've never really done much unit-testing in an object-oriented language.

Anyway. I don't like the idea of having to prefix the interfaces with I, nor do I like the idea of having to write `impl` or anything of the sort at the end. I feel like an interface should be different enough from an implementation that there should not even be any naming conflicts (to me, an interface simply describes a few elements of behaviour and should not describe the whole of a class; what would even be the point of the interface if it is _that_ specific?). 

But I gather that in order to write some unit-tests, we need to "force" the use of interfaces. In this case, I would choose the lesser evil and go with the prefix I, and lose the `impl`.

Reply:
Aight I'll change it. I choosed the Impl way since Impl is only written on definition and once on initialization, then we pass it as the interface. The I prefix need to be everywhere, so you see it "more often". 

Anyway, do we need small confluence page for some conventions like that or it won't even be used? (file naming, interface naming, abstract class naming, etc).

Reply:
I think a confluence page would be helpful. Settle these conversations outside of a review to concentrate on actual code and less on its looks :smile: